real estate module (ibmah)

ibmah-odoo opened the pull request:
closes #1062. This one is based on master instead of v19.0

ibmah-odoo committed:
[ADD] estate: introduce structure for estate module

Chapter 2

ibmah-odoo committed:
[IMP] estate: Add estate_property model

Chapter 3

ibmah-odoo committed:
[IMP] estate: Add access right rules to estate module data

Chapter 4

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1069.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1069)

amah-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,15 @@
+    'name': "estate",
+    'version': '1.0',
+    'depends': ['base'],
+    'author': "ibrahim (ibmah)",
```
The author is always `'Odoo S.A.'` for the RD because to develop for odoo it is not out own apps 🤔

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,26 @@
+                   ('east', 'East'),
+                   ('west', 'West')],
+        help = 'Indicates the direction the garden is facing with respect to the house')
+
```
we tend to leave only 1 line at the end of the file

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,26 @@
+        selection=[('north', 'North'),
+                   ('south', 'South'),
+                   ('east', 'East'),
+                   ('west', 'West')],
```
A personal pref but want to show another way to structure this
```suggestion
    garden_orientation = fields.Selection([
        ('north', 'North'),
        ('south', 'South'),
        ('east', 'East'),
        ('west', 'West')
    ], help = "put it here")
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,26 @@
+class EstateProperty(models.Model):
+    _name = "estate_property"
+    _description = "Estate Property data model"
+    name = fields.Char(required=True)
```
we can have a blank line to make clear the end of the metadata and fields

amah-odoo reviewed (commented):
Let's go with the first review 🎉

ibmah-odoo committed:
[IMP] estate: Implemet menus and new model customizations

Chapter 5

amah-odoo commented on .gitignore:
```diff
@@ -127,3 +127,4 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+pyrightconfig.json
```
do not add your own `.gitignore` to the commits

amah-odoo reviewed (commented):
This is mainly about the commit msg let's try working to have the chapter number in the commit header (commit first line) and I want you to add some description for each commit body with what is done in this chapter

ibmah-odoo committed:
[IMP] estate: Add list, form, search views to real estate module

Chapter 6

ibmah-odoo committed:
[IMP] estate: Add relations between real estate property models

Chapter 7

This adds tags to properties, lists offers made (whether refued or
accepted) along with the salesperson and buyer in "Other info" tab

ibmah-odoo committed:
[IMP] estate: introduce computed fields and use onchange effects for updates

Chapter 8

ibmah-odoo committed:
[IMP] estate: implement actions on properties and offers

Chapter 9

This introduces marking a property as "Sold" or "Cancelled".

It also allows for easier offers management from the list view and it
gets reflected automatically on the property

ibmah-odoo committed:
[IMP] estate: add constraints on property and offer prices as well as tags and types

Chapter 10

This commit adds checks on property and offer prices to ensure they are
positive and that no property sells for less than 90% of the expected
price. It also ensures the uniqueness of property tags and types.

amah-odoo commented on estate/models/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from . import (estate_property, estate_property_offer, estate_property_tag,
+               estate_property_type)
```
we tend to use this structure in the imports
```suggestion
from . import estate_property
from . import estate_property_offer
from . import estate_property_tag
from . import estate_property_type
```

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,116 @@
+        copy=False, default=datetime.now() + relativedelta(month=3)
+    )
+    expected_price = fields.Float(required=True)
+    _check_expected_price = models.Constraint(
```
can you move all the `models.Constraint` after all the fields are defined

amah-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,116 @@
+            if record.state != "sold":
+                record.state = "cancelled"
+            else:
+                raise exceptions.UserError("can't cancel a sold property")
```
it is not optimal to import all the module to have only 1 or 2 object so we use to
`from odoo.exceptions import UserError` and use it directly 
```suggestion
                raise UserError("can't cancel a sold property")
```

ibmah-odoo replied:
I usually prefer qualified imports since they prevent namespace pollution and name clashes on larger modules at the cost of being a tiny bit verbose.

It's a common misconception that importing a single object from a module doesn't cause python to import and cache the entire module. There's no performance hit for importing a module compared to an object in the module.

> The only difference between the two statements is what name is bound

[ref](https://softwareengineering.stackexchange.com/a/187471)

ibmah-odoo replied:
If that's the convention in odoo codebase, I'd happily adopt it but thought I'd put this info out there

amah-odoo reviewed (commented):
Small comments about .py structure you can refer to this [DOC](https://www.odoo.com/documentation/18.0/contributing/development/coding_guidelines.html#symbols-and-conventions) to know about it all

ibmah-odoo committed:
[IMP] estate: implement inline views, widgets and list ordering of models

Chapter 11:

Implement inlinen view for properties by their type, add statusbar and
tag color picker widgets, implement stat button for displaying the
number of offers for each property type

ibmah-odoo committed:
[IMP] estate: inherit models and views from res.users

Chapter 12
This overrides deletion methods to make them safer by only allowing deletion of new or cancelled properties.
It integrates the real estate module with the views for 'res.users' model by adding a new page listing the properties managed by each user

ibmah-odoo committed:
[IMP] estate: link real estate module with accounting to auto generate invoices on sales

Chapter 13

This introduces a link module "estate_account" that enables real estate
module to generate invoices automatically on sold houses.

ibmah-odoo committed:
[IMP] estate: implement kanban view

Chapter 14

This implements kanban views using QWeb template definition

ibmah-odoo committed:
[CLN] estate: address warnings from missing licenses and model names

Chapter 15

This addresses warnings and various runbot-reported issues.
